pkg/images/images.go

```diff
@@ -0,0 +1,152 @@
+package images
+
+import (
+	"fmt"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/utils"
+	corev1 "k8s.io/api/core/v1"
+)
+
+const DefaultRegistry = "docker.io"
+
+// ImageId is the set of components that uniquely identify a container image: registry, repository, name and tag.
+type ImageId interface {
+	GetRegistry() string
+	GetRepository() string
+	GetName() string
+	GetTag() string
+}
+
+type imageId struct {
+	registry   string
+	repository string
+	name       string
+	tag        string
+}
+
+func (i *imageId) GetRegistry() string {
+	return i.registry
+}
+
+func (i *imageId) GetRepository() string {
+	return i.repository
+}
+
+func (i *imageId) GetName() string {
+	return i.name
+}
+
+func (i *imageId) GetTag() string {
+	return i.tag
+}
+
+// NewImageId returns a new ImageId for the given components. It does not check that the components are valid.
+func NewImageId(registry, repository, name, tag string) ImageId {
+	return &imageId{
+		registry:   registry,
+		repository: repository,
+		name:       name,
+		tag:        tag,
+	}
+}
+
+// Image uniquely identifies a container image and specifies how to pull it from its remote registry.
+type Image interface {
+	ImageId
+	GetPullPolicy() corev1.PullPolicy
+	GetPullSecretRef() *corev1.LocalObjectReference
+}
+
+type image struct {
+	ImageId
+	pullPolicy    corev1.PullPolicy
+	pullSecretRef *corev1.LocalObjectReference
+}
+
+func (i *image) GetPullPolicy() corev1.PullPolicy {
+	return i.pullPolicy
+}
+
+func (i *image) GetPullSecretRef() *corev1.LocalObjectReference {
+	return i.pullSecretRef
+}
+
+// NewImage returns a new Image for the given id and pull settings. It panics if id is nil. It does not check that the
+// image components are valid.
+func NewImage(id ImageId, pullPolicy corev1.PullPolicy, pullSecretRef *corev1.LocalObjectReference) Image {
+	return &image{
+		ImageId:       id,
+		pullPolicy:    pullPolicy,
+		pullSecretRef: pullSecretRef,
+	}
+}
+
+// ImageString returns the string that uniquely identifies the given ImageId. It panics if id is nil. It does not check
+// if the id components are valid.
+func ImageString(i ImageId) string {
```

Comment:
`Image` could inherit from `fmt.Stringer` but that would force all structs satisfying `Image` to also implement `String()` – since there are no default methods in Go.